File: #26-50 (Ada, C++, FORTRAN)/C++/Euler37.cpp
/////////////////////////////////////////////////////////////////////
// Euler37 ~ Truncatable primes
// 
// Problem Statement:
// 
// The number 3797 has an interesting property. Being prime itself,
// it is possible to continuously remove digits from left to right,
// and remain prime at each stage: 3797, 797, 97, and 7. Similarly
// we can work from right to left: 3797, 379, 37, and 3.
// 
// Find the sum of the only eleven primes that are both truncatable
// from left to right and right to left.
// 
// NOTE: 2, 3, 5, and 7 are not considered to be truncatable primes.
//
/////////////////////////////////////////////////////////////////////

/////////////////////////////////////////////////////////////////////
// Algorithm Explanation:
// 
// Iterates upwards checking if the number truncates from the left
// and right side while maintaining primality. Uses recursion to
// accept arbitrary length integers. When found the integers are
// added to the solution.
//
/////////////////////////////////////////////////////////////////////

#include <iostream>
#include <string>
#include <chrono>

using namespace std;

bool prime(int num)
{
	for (int i = 2; i <= sqrt(num); i++)
	{
		if (num % i == 0)
		{
			return 0;
		}
	}
	if (num == 1)
	{
		return 0;
	}
	return 1;
}

bool truncatable_left(int num)
{
	if (prime(num))
	{
		string s1 = to_string(num);

		if (s1.length() > 2)
		{
			s1 = s1.substr(1, s1.length() - 1);
		}
		else
		{
			s1 = s1.substr(1, 1);
		}
		num = stoi(s1, nullptr, 0);
		
		if (num > 10)
		{
			return truncatable_left(num);
		}
		else
		{
			if (prime(num))
			{
				return 1;
			}
		}
	}
	return 0;
}

bool truncatable_right(int num)
{
	if (prime(num))
	{
		string s1 = to_string(num);

		if (s1.length() > 2)
		{
			s1 = s1.substr(0, s1.length() - 1);
		}
		else
		{
			s1 = s1.substr(0, 1);
		}
		num = stoi(s1, nullptr, 0);

		if (num > 10)
		{
			return truncatable_right(num);
		}
		else
		{
			if (prime(num))
			{
				return 1;
			}
		}
	}
	return 0;
}

int main()
{
	auto start = std::chrono::high_resolution_clock::now();
	int solution = 0;

	int i = 10;
	int primes = 0;
	while (primes < 11)
	{
		if (truncatable_left(i))
		{
			if (truncatable_right(i))
			{
				primes++;
				solution += i;
			}
		}
		i++;
	}

	auto done = std::chrono::high_resolution_clock::now();
	cout << "Solution: " << solution << endl << endl;
	cout << "Program completed in " << std::chrono::duration_cast<std::chrono::milliseconds>(done - start).count() << " milliseconds." << endl;
}

File: #26-50 (Ada, C++, FORTRAN)/FORTRAN/Euler37.f90
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Euler37 ~ Truncatable primes
! 
! Problem Statement:
! 
! The number 3797 has an interesting property. Being prime itself,
! it is possible to continuously remove digits from left to right,
! and remain prime at each stage: 3797, 797, 97, and 7. Similarly
! we can work from right to left: 3797, 379, 37, and 3.
! 
! Find the sum of the only eleven primes that are both truncatable
! from left to right and right to left.
! 
! NOTE: 2, 3, 5, and 7 are not considered to be truncatable primes.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Algorithm Explanation:
! 
! Iterates upwards checking if the number truncates from the left
! and right side while maintaining primality. Uses recursion to
! accept arbitrary length integers. When found the integers are
! added to the solution.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

program Euler37
	
	implicit none
	integer :: time_start, time_finish, clock_finish, solution = 0
	integer :: i = 10, primes = 0
	call SYSTEM_CLOCK(time_start)
	do while(primes < 11)
		if (truncatable_left(i)) then
			if(truncatable_right(i)) then
				primes = primes + 1
				solution = solution + i
			end if
		end if
		i = i + 1
	end do

	call SYSTEM_CLOCK(time_finish)
	clock_finish = time_finish - time_start
	write (*,"(A10,I10,//,A21,I5,A14)") "Solution: ",solution,"Program completed in ",clock_finish," milliseconds."

contains

	logical function prime(num)
		integer :: num, i
		if (num == 1) then
			prime = .false.
			return
		end if
		do i = 2,INT(sqrt(real(num)))
			if(modulo(num,i) == 0) then
				prime = .false.
				return
			end if
		end do
		prime = .true.
		return
    end
	
	recursive function truncatable_left(num) result(r)
		implicit none
		integer, intent(in) :: num
		integer :: n
		character(10) :: s1
		character(:), allocatable :: str
		logical :: r
		n = num
		if (prime(num)) then
			write (s1,'(I0)') n
			str = trim(s1)
			if (LEN(str) > 2) then
				str = trim(str(2:LEN(str)))
			else
				str = str(2:2)
			end if
			read(str,*) n
		
			if (n > 10) then
				r = truncatable_left(n)
				return
			else
				if (prime(n)) then
					r = .true.
					return
				end if
			end if
		end if
		r = .false.
		return
	end
	
	recursive function truncatable_right(num) result(r)
		implicit none
		integer, intent(in) :: num
		integer :: n
		character(10) :: s1
		character(:), allocatable :: str
		logical :: r
		n = num
		if(prime(num)) then
			write(s1,'(I0)') n
			str = trim(s1)
			if (LEN(str) > 2) then
				str = str(1:LEN(str)-1)
			else
				str = trim(str(1:1))
			end if
			read(str,*) n
			if (n > 10) then
				r = truncatable_right(n)
				return
			else
				if (prime(n)) then
					r = .true.
					return
				end if
			end if
		end if
		r = .false.
		return
	end

end program Euler37
